exercises/book-store/tests/book-store.rs

```diff
@@ -0,0 +1,135 @@
+//! Tests for book-store
+//!
+//! Generated by [script][script] using [canonical data][canonical-data]
+//!
+//! [script]: https://github.com/exercism/rust/blob/master/bin/init_exercise.py
+//! [canonical-data]: https://raw.githubusercontent.com/exercism/problem-specifications/master/exercises/book-store/canonical_data.json
+
+extern crate book_store;
+use book_store::*;
+
+/// Process a single test case for the property `total`
+///
+/// All cases for the `total` property are implemented
+/// in terms of this function.
+///
+/// Note that you'll need to both name the expected transform which
+/// the student needs to write, and name the types of the inputs and outputs.
+/// While rustc _may_ be able to handle things properly given a working example,
+/// students will face confusing errors if the `I` and `O` types are not concrete.
+///
+/// Expected input format: ('basket', 'targetgrouping')
```

Comment:
Should it be the case that the student still sees this? They may be confused because basket and targetgrouping are nowhere in sight here

Reply:
This is a note to the exercise implementor, but in this case I think we should leave it for clarity for those students who examine the tests.

On test generation, the generator collects the cases. If the case doesn't already have an `input` property, it generates one as a tuple containing all the unknown properties in the case. However, this discards information: as properties, these items have names; as tuple fields, they do not. The comment is inserted to help the implementer decipher the input tuple.

Reply:
I was about to suggest

> Ah, then this should be `process_total_case(basket: Vec<usize>, targetgrouping: Vec<Vec<usize>>, expected: f64)`

but now I hesitate a little bit about that suggestion because then it might not be clear which of these are inputs, whereas it is clar if they're named `input:`

It's also unfortuntae because `targetgrouping` is not an input. Maybe it shouldn't be in the canonical-data.json at all.

Reply:
but we can continue this discussion pretending it's an input, based on what we would like to see for exercises that actually do have multiple inputs.

Reply:
The principle I used when writing the generator is that in cases where we require manual intervention from the test implementor, the effort required should be minimal. The actual generated code looks like this:

```rust
fn process_{property}_case<I, O>(input: I, expected: O) {
    // typical implementation:
    // assert_eq!(
    //     student_{property}_func(input),
    //     expected
    // )
    unimplemented!()
}
```

While it would be possible to break that out into a more complex function signature, as you say, it becomes less obvious for the test implementer which parameters form the actual inputs. I prefer to leave them as a simple tuple, for now.